libs/s25main/buildings/nobMilitary.cpp

```diff
@@ -81,6 +82,22 @@ size_t nobMilitary::GetTotalSoldiers() const
     return sum;
 }
 
+std::array<unsigned, NUM_SOLDIER_RANKS> nobMilitary::GetTotalSoldiersByRank() const
+{
+    std::array<unsigned, NUM_SOLDIER_RANKS> counts = {0};
+    for(auto& troop : troops)
+        ++counts[troop->GetRank()];
+    for(auto& troop : ordered_troops)
+        ++counts[troop->GetRank()];
+    for(auto& troop : troops_on_mission)
+        ++counts[troop->GetRank()];
+    if(defender_ && (defender_->IsWaitingAtFlag() || defender_->IsFightingAtFlag()))
+        ++counts[defender_->GetRank()];
+    for(auto& troop : far_away_capturers)
+        ++counts[troop->GetRank()];
+    return counts;
```

Comment:
Not fully sure here. Are you sure that the additional conditions for `defender_` are required and it is not in `troops_on_mission` or `far_away_capturers`?

Just double checking.

Reply:
Fairly sure, `defender_` always comes out of the building(`troops`) from which he is removed in `ProvideDefender` and not added anywhere else.

Reply:
Ok, but IIRC as soon as he is removed from `troops` it is set as `defender_`, so I don't understand the ` && (defender_->IsWaitingAtFlag() || defender_->IsFightingAtFlag())` condition part. Wouldn't that leave a (short time) where the defender is wrongly not counted?

Reply:
That's to match the number returned by `GetTotalSoldiers` which calculates it like this. I'm struggling to figure out why it does that though, I don't understand it either.

Reply:
See here: https://github.com/Return-To-The-Roots/s25client/commit/a8fbe91a23edd08151943f43f3e6135fbf5bae9b